### DISCUSSION TITLE
vsgraytracing demo   add shadow.rmiss

### DISCUSSION THREAD

**Discussion by WdsPjm** [vsg-dev/VulkanSceneGraph#977] [Q&A]
vsgraytracing demo, I want to add shadow.rmiss, but it hasn't been connected. Here's my code. I don't know how to connect shadow.rmiss in traceRays.

 // fallback to spv shaders if GLSL variants can't be loaded
        auto  raygenShader = vsg::ShaderStage::read(VK_SHADER_STAGE_RAYGEN_BIT_KHR, "main", "shaders/raygen.rgen.spv", options);
        auto missShader = vsg::ShaderStage::read(VK_SHADER_STAGE_MISS_BIT_KHR, "main", "shaders/miss.rmiss.spv", options);
        auto shadowthitShader = vsg::ShaderStage::read(VK_SHADER_STAGE_MISS_BIT_KHR, "main", "shaders/shadow.rmiss.spv", options);
        auto closesthitShader = vsg::ShaderStage::read(VK_SHADER_STAGE_CLOSEST_HIT_BIT_KHR, "main", "shaders/closesthit.rchit.spv", options);

        if (!raygenShader || !missShader || !closesthitShader|| !shadowthitShader)
        {
            std::cout << "Could not create shaders." << std::endl;
            return 1;
        }

        auto shaderStages = vsg::ShaderStages{raygenShader, missShader,shadowthitShader, closesthitShader};

        // set up shader groups
        auto raygenShaderGroup = vsg::RayTracingShaderGroup::create();
        raygenShaderGroup->type = VK_RAY_TRACING_SHADER_GROUP_TYPE_GENERAL_KHR;
        raygenShaderGroup->generalShader = 0;

        auto missShaderGroup = vsg::RayTracingShaderGroup::create();
        missShaderGroup->type = VK_RAY_TRACING_SHADER_GROUP_TYPE_GENERAL_KHR;
        missShaderGroup->generalShader = 1;

        auto shadowShaderGroup = vsg::RayTracingShaderGroup::create();
        shadowShaderGroup->type = VK_RAY_TRACING_SHADER_GROUP_TYPE_GENERAL_KHR;
        shadowShaderGroup->generalShader = 2;
        

        auto closestHitShaderGroup = vsg::RayTracingShaderGroup::create();
        closestHitShaderGroup->type = VK_RAY_TRACING_SHADER_GROUP_TYPE_TRIANGLES_HIT_GROUP_KHR;
        closestHitShaderGroup->closestHitShader = 3;

        auto shaderGroups = vsg::RayTracingShaderGroups{raygenShaderGroup, missShaderGroup,shadowShaderGroup, closestHitShaderGroup};

        auto traceRays = vsg::TraceRays::create();
        traceRays->raygen = raygenShaderGroup;
        traceRays->missShader = missShaderGroup;
        // add shadow.rmiss  ???
        traceRays->hitShader = closestHitShaderGroup;
        traceRays->width = windowTraits->width;
        traceRays->height = windowTraits->height;
        traceRays->depth = 1;

**Comment by robertosfield** [comment]
vsg::TraceRays just encapsulates vkCmdTraceRays:

https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdTraceRaysKHR.html

This doesn't have a dedicate shadow shader group, if this is what you were expecting.  I don't know what your expectations and what specifically you trying to do in your code and why so I can't say what you should be doing.

If you are trying to follow a 3rd party Viulkan raytracing examplet and are trying to implement this with the VSG then provide a link to the example/tutorial so we can see where you are coming from and see if we can spot what you should do.

**Comment by WdsPjm** [comment]
I refer to https://github.com/SaschaWillems/Vulkan/blob/master/examples/raytracingshadows/raytracingshadows.cpp

There is a ShaderBindingTable, and the code is as follows
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/52599568/933ff716-7ce5-4ade-8094-afe5879c60cd)


However, there is no processing in vsg for two Missshaders (miss.rmiss shadow.rmiss). I ran the following tests in vsg. Then it worked
![image](https://github.com/vsg-dev/VulkanSceneGraph/assets/52599568/975954a9-e25e-4c60-8c3f-ba1fc1471119)

**Reply by robertosfield** [reply]
I am afraid can't follow what your changes are meant to do and how they might work, having special handling for i==1 just looks wrong.  I've looked at the original RayTracingPipeline.cpp and it looks more sensible than your changes.

FYI, I'm not the author of original RTX ray tracing classes, nor their subsequent updated to use Khronos ray tracing functions, so I'm mainly just looking from a general C++ point of view.  At some point in the future I'll have time to properly work with ray tracing but unfortunately for now I'm playing catch up.
